### ISSUE TITLE
Add Memory / Storage Metrics

### ISSUE THREAD

**Issue by vmorris** [zhmcclient/zhmc-prometheus-exporter#112]
Is it possible to add cec overall and per partition memory usage metrics?

**Comment by andy-maier** [comment]
Hi Vance. Sorry for the delay.

Since 0.6.0, the exporter supports all metrics the HMC provides. The HMC does not provide any memory usage metrics at all. Since memory cannot be overcommitted, I think the idea is that providing the allocated memory is sufficient, and the memory usage would need to come from the OS running in each partition.

The allocated memory at the CPC level and at the partition level is provided as properties of the respective resource objects, but not as metrics. Since the zhmc exporter only supports the exporting of items the HMC provide at its metrics interface, these items do not include the memory allocation values. The zhmc exporter could be extended to provide artificial metrics about certain allocation values, e.g. memory or CPU. Let me know if you think that would be useful to have in Prometheus.

**Comment by vmorris** [comment]
@andy-maier Thank you for the explanation! I would find this very useful to see the memory allocations per partition.
Regarding CPU, I also think it would be good to collect the CPU allocations and weights, at least.

**Comment by andy-maier** [comment]
Ok. Sounds good to me.

Here are the CPU and memory properties that could be reported as Prometheus metrics:

CPC level:
* CPC.processor-count-{cpu-type}
* CPC.storage-total-installed
* CPC.storage-hardware-system-area
* CPC.storage-customer
* CPC.storage-customer-central
* CPC.storage-customer-expanded
* CPC.storage-customer-available

Partitions on CPCs in DPM mode:
* Partition.{cpu-type}-processors
* Partition.{cpu-type}-absolute-processor-capping (maybe not, since it can be concluded from the -value)
* Partition.{cpu-type}-absolute-processor-capping-value
* Partition.{cpu-type}-processing-weight-capped
* Partition.minimum-{cpu-type}-processing-weight
* Partition.initial-{cpu-type}-processing-weight
* Partition.current-{cpu-type}-processing-weight
* Partition.maximum-{cpu-type}-processing-weight
* Partition.threads-per-processor
* Partition.initial-memory
* Partition.reserved-memory
* Partition.maximum-memory

There are similar (but slightly different) properties for LPARs on CPCs in classic mode.

**Comment by andy-maier** [comment]
I have opened https://github.com/zhmcclient/python-zhmcclient/issues/762 to implement the resource caching.

**Comment by andy-maier** [comment]
zhmcclient PR https://github.com/zhmcclient/python-zhmcclient/pull/814 adds support for auto-updatied resource objects.

I have started PR #158 to implement resource property based metrics using that new auto-update support.

**Comment by andy-maier** [comment]
@vmorris Vance, PR #158 is ready to be tried out. It adds support for several CPC and partition metrics around processors and memory.

**Comment by andy-maier** [comment]
PR #158 has been merged to master and will be in the upcoming 0.8.0 version.

**Comment by vmorris** [comment]
> @vmorris Vance, PR #158 is ready to be tried out. It adds support for several CPC and partition metrics around processors and memory.

I tried testing off master branch today and met the following:

``` 
$ zhmc_prometheus_exporter -c hmccreds.yaml -m metrics.yaml
Error: Error returned from HMC at z15hmc-01.mgmt.upshift.eng.rdu2.redhat.com: 400,7: The following supplied metric group name(s) are invalid: environmental-power-status [POST /api/services/metrics/context]
```


I set `environmental-power-status.fetch` to false and then:


```
$ zhmc_prometheus_exporter -c hmccreds.yaml -m metrics.yaml
Traceback (most recent call last):
  File "/home/fedora/venv-zhmc-prometheus-exporter/bin/zhmc_prometheus_exporter", line 33, in <module>
    sys.exit(load_entry_point('zhmc-prometheus-exporter==0.8.0.dev1', 'console_scripts', 'zhmc_prometheus_exporter')())
  File "/home/fedora/venv-zhmc-prometheus-exporter/lib64/python3.9/site-packages/zhmc_prometheus_exporter/zhmc_prometheus_exporter.py", line 1140, in main
    context, resources = create_metrics_context(
  File "/home/fedora/venv-zhmc-prometheus-exporter/lib64/python3.9/site-packages/zhmc_prometheus_exporter/zhmc_prometheus_exporter.py", line 548, in create_metrics_context
    lpar.enable_auto_update()
AttributeError: 'Lpar' object has no attribute 'enable_auto_update'
```

Any guidance? Thanks!

**Comment by andy-maier** [comment]
Since the PR was already merged, I created issue #170.
